src/eager/tensor_handle.rs

```diff
@@ -0,0 +1,273 @@
+            if status.is_ok() {
+                Ok(CStr::from_ptr(device_name).to_str()?.to_string())
+            } else {
+                Err(status)
+            }
+        }
+    }
+
+    /// Returns the name of the device in whose memory `h` resides.
+    /// This function will block till the operation that produces `h` has completed.
+    pub fn backing_device_name(&self) -> Result<String> {
+        let status = Status::new();
+        unsafe {
+            let device_name = tf::TFE_TensorHandleBackingDeviceName(self.inner, status.inner);
+            if status.is_ok() {
+                Ok(CStr::from_ptr(device_name).to_str()?.to_string())
+            } else {
+                Err(status)
+            }
+        }
+    }
+
+    /// Return a new TensorHandle that shares the underlying tensor with `h`.
+    pub fn copy_sharing_tensor(&self) -> Result<Self> {
+        let status = Status::new();
+        let inner = unsafe { tf::TFE_TensorHandleCopySharingTensor(self.inner, status.inner) };
+        if status.is_ok() {
+            Ok(Self {
+                inner,
+                ctx: self.ctx,
+            })
+        } else {
+            Err(status)
+        }
+    }
+
+    /// This function will block till the operation that produces `h` has
+    /// completed. The memory returned might alias the internal memory used by
+    /// TensorFlow. Hence, callers should not mutate this memory.
+    pub fn resolve<T: TensorType>(&self) -> Result<Tensor<T>> {
+        let mut status = Status::new();
+        let tf_tensor = unsafe { tf::TFE_TensorHandleResolve(self.inner, status.inner) };
+        if !status.is_ok() {
+            return Err(status);
+        }
+        if self.data_type() != T::data_type() {
+            let msg = format!(
+                "The expected data type ({}) and underlying data type ({}) did not match.",
+                T::data_type(),
+                self.data_type()
+            );
+
+            status.set_lossy(crate::Code::InvalidArgument, &msg);
+            return Err(status);
+        }
+
+        // Safely unwrap since data_type was checked beforehand.
+        unsafe { Ok(Tensor::from_tf_tensor(tf_tensor).unwrap()) }
+    }
+
+    /// Create a new TensorHandle with the same contents as 'h' but placed
+    /// in the memory of the device name 'device_name'.
+    /// If source and destination are the same device, then this creates a new handle
+    /// that shares the underlying buffer. Otherwise, it currently requires at least
+    /// one of the source or destination devices to be CPU (i.e., for the source or
+    /// destination tensor to be placed in host memory).
+    /// If async execution is enabled, the copy may be enqueued and the call will
+    /// return "non-ready" handle. Else, this function returns after the copy has
+    /// been done.
+    pub fn copy_to_device(&self, ctx: &Context, device_name: &str) -> Result<Self> {
```

Comment:
What I intended to do was as you said, but the implementation did not seem to be so (it was bounded by the same lifetime of `self`).

In order to associate the lifetime of TensorHandle with `ctx`, would it be better to implement it as another associated function as follows?

```rust
impl<'a, 'b> TensorHandle<'b> {
    pub fn copy_to_device(
        h: &'a TensorHandle,
        ctx: &'b Context,
        device_name: &str,
    ) -> Result<Self> {
        let status = Status::new();

        let device_name = CString::new(device_name)?;
        unsafe {
            let inner = tf::TFE_TensorHandleCopyToDevice(
                h.inner,
                ctx.inner,
                device_name.as_ptr(),
                status.inner,
            );

            if status.is_ok() {
                Ok(Self {
                    inner,
                    ctx: PhantomData,
                })
            } else {
                Err(status)
            }
        }
    }
}
```

Reply:
I modified the code so that the lifetime of the created TensorHandle is bounded by input context. Could you review again if the implementation is as intended?